### PR TITLE
[Playwright] PEPPER-1281 dss view permissions test

### PR DESCRIPTION
**Pull request by aweng98**
[PEPPER-1281](https://broadworkbench.atlassian.net/browse/PEPPER-1281)

New users to be added to the vault secret file after PR merge. Organized DSM users in vault file in chronological order. 
Screenshot 2024-01-25 at 5 22 49 PM

[PEPPER-1281]: https://broadworkbench.atlassian.net/browse/PEPPER-1281?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Note:

> In Participant List page, only columns that relates to the tabs they can see should be visible.

This check is skipped because I'm not certain if currently visible columns per test user with limited permssions are correctly showing.

Additional changes to address flaky areas after ran all tests several times on localhost.

### CODE REVIEW

File: playwright-e2e/README.md [hunk 1]
```diff
@@ -44,6 +44,39 @@ Worthy of note:
 If you don't want to use the `.env` file, you can also specify environment variables in cmd. See an example of this in **Examples** section.
 
 
+## Test Users
```
Comment (aweng98) [inline playwright-e2e/README.md:47]: Show users and permissions in screenshots. 

Screenshot 2024-01-25 at 5 24 57 PM

File: playwright-e2e/dsm/pages/participant-list-page.ts [hunk 2]
```diff
@@ -405,12 +405,13 @@ export default class ParticipantListPage extends DsmPageBase {
...
   async findParticipantFor(columnGroup: string, columnName: string, opts: {value?: string, nth?: number} = {}): Promise<number> {
-    const { value, nth = 1 } = opts;
+    const { value, nth = 0 } = opts;
 
     const compareForMatch = async (index: number): Promise<number> => {
+      console.log('\n\n');
       const columnText = await participantListTable.getTextAt(index, columnName);
       if (value) {
-        return columnText.some(text => text.indexOf(value) !== -1) ? index : -1;
+        return columnText.some(text => text.trim().length > 0 && new RegExp(value, 'i').test(text.trim())) ? index : -1;
```
Comment (aweng98) [inline playwright-e2e/dsm/pages/participant-list-page.ts:414]: Changed to case-insensitive matching.

File: playwright-e2e/tests/dsm/miscellaneous/permissions/view-seq-order-permissions.spec.ts [hunk 3]
```diff
@@ -16,82 +16,82 @@ import Select from 'dss/component/select';
 import { testLimitedPermissions as test } from 'fixtures/dsm-fixture';
 
 test.describe('View Sequencing Order Permission Test', () => {
-    const studies = [StudyEnum.OSTEO2, StudyEnum.LMS]; //Current clinical studies
-    const testEmail = process.env.DSM_USER1_EMAIL as string;
+  const studies = [StudyEnum.OSTEO2, StudyEnum.LMS]; //Current clinical studies
```
Comment (aweng98) [inline playwright-e2e/tests/dsm/miscellaneous/permissions/view-seq-order-permissions.spec.ts:19]: Fix formatting

File: playwright-e2e/tests/dsm/osteo/os1-participant-p4a42b.spec.ts [hunk 4]
```diff
@@ -74,11 +73,6 @@ test.describe.serial('Same Participant in Osteo1 and Osteo2', () => {
         .toHaveScreenshot('osteo1-medical-release-form-expanded-view.png');
       await surveyTabLocator.locator('mat-expansion-panel-header').nth(2).click(); // collapse click
 
-      // Compare Sample Information tab screenshot
```
Comment (aweng98) [inline playwright-e2e/tests/dsm/osteo/os1-participant-p4a42b.spec.ts:77]: UI displayed data is not consistent.

File: playwright-e2e/dsm/pages/miscellaneous-pages/user-and-permissions-page.ts [hunk 5]
```diff
@@ -14,10 +14,12 @@ export default class UserPermissionPage {
   private DARWIN_ARK_STUDY_GROUP = 'darwin';
   private currentStudyPermissions: UserPermission[];
   private readonly page: Page;
+  private readonly root: Locator;
 
   constructor(page: Page) {
     this.page = page;
     this.currentStudyPermissions = [];
+    this.root = this.page.locator('app-users-and-permissions');
```
Comment (aweng98) [inline playwright-e2e/dsm/pages/miscellaneous-pages/user-and-permissions-page.ts:22]: Xpath updated in this file.

File: playwright-e2e/.env.sample [hunk 6]
```diff
@@ -15,17 +15,27 @@ ATCP_AUTH0_CLIENT_SECRET=
...
 # fourth user with very limited kit_shipping permissions
 DSM_USER4_EMAIL=
 DSM_USER4_PASSWORD=
+# fifth user for mr-view-permission.spec
+DSM_USER5_EMAIL=
+DSM_USER5_PASSWORD=
+# sixth user: dss-view-permission.spec
+DSM_USER6_EMAIL=
+DSM_USER6_PASSWORD=
+# seventh user: pt-list-view-permission.spec
```
Comment (Rinchan20) [inline playwright-e2e/.env.sample:35]: good call making sure to note which user is for which test

File: playwright-e2e/authentication/auth-base.ts [hunk 7]
```diff
@@ -14,30 +15,51 @@ export async function fillInEmailPassword(
 ): Promise<void> {
   const { email, password, waitForNavigation = true, waitForAuth = true } = opts;
 
-  await Promise.race<boolean>([
-    page.locator('#signUpForm').isVisible(),
-    page.locator('.auth0-lock-header-welcome').isVisible()
+  const emailInput = page.locator('input[type="email"]').locator('visible=true');
```
Comment (aweng98) [inline playwright-e2e/authentication/auth-base.ts:18]: refactor func in this file as old code is not working in all situations.

File: playwright-e2e/authentication/auth-dsm.ts [hunk 8]
```diff
@@ -25,20 +15,38 @@ export async function login(page: Page, opts: { email?: string; password?: strin
...
-    await waitForNoSpinner(page);
-    await page.waitForSelector('.auth0-lock-header-welcome', { state: 'visible', timeout: 2 * 1000 })
-      .then(async () => {
-        await fillInEmailPassword(page, { email, password, waitForNavigation: false });
-        await assertLoggedIn(page);
-      })
-      .catch((error) => {
-        throw new Error(error); // stops test
-      });
+  const authWindow = page.locator('.auth0-lock-header-welcome');
```
Comment (aweng98) [inline playwright-e2e/authentication/auth-dsm.ts:18]: refactoring code.

File: playwright-e2e/dsm/pages/dsm-page-base.ts [hunk 9]
```diff
@@ -11,11 +12,14 @@ export default abstract class DsmPageBase {
   }
 
   async reload(): Promise<void> {
-    const networkIdlePromise = this.page.waitForLoadState('networkidle');
-    await this.page.reload();
-    await this.page.waitForLoadState('load');
-    await networkIdlePromise;
+    const logRequest = (interceptedRequest: Request) => {
```
Comment (aweng98) [inline playwright-e2e/dsm/pages/dsm-page-base.ts:15]: monitor for any failed request during reload.

File: playwright-e2e/dsm/pages/kitUpload-page/kitUpload-page.ts [hunk 10]
```diff
@@ -8,30 +7,24 @@ import {StudyEnum} from 'dsm/component/navigation/enums/selectStudyNav-enum';
 import path from 'path';
 import Modal from 'dsm/component/modal';
 import {logInfo} from 'utils/log-utils';
+import KitsPageBase from 'dsm/pages/kits-page-base';
 
-export default class KitUploadPage {
-  private readonly PAGE_TITLE = 'Kit Upload';
+export default class KitUploadPage extends KitsPageBase {
```
Comment (aweng98) [inline playwright-e2e/dsm/pages/kitUpload-page/kitUpload-page.ts:12]: extends `KitsPageBase` to eliminate duplicated code.

File: playwright-e2e/dsm/pages/kitsInfo-pages/kitsSentPage.ts [hunk 11]
```diff
@@ -25,26 +25,21 @@ export default class KitsSentPage extends KitsPageBase {
     await this.kitsTable.rowsPerPage(rows);
   }
 
-  public async waitForLoad(): Promise<void> {
```
Comment (aweng98) [inline playwright-e2e/dsm/pages/kitsInfo-pages/kitsSentPage.ts:28]: `waitForLoad` is duplicate of `waitForReady` in `KitsPageBase.ts`

File: playwright-e2e/fixtures/fixture-base.ts [hunk 12]
```diff
@@ -18,6 +19,12 @@ export const fixtureBase = base.extend({
     await page.route('**/*', async (route) => {
       return REQUEST_EXCLUDES.some((urlPart) => route.request().url().includes(urlPart)) ? route.abort() : route.continue();
     });
+    // Listen for all console events and log errors to the Terminal window
+    page.on('console', msg => {
+      if (msg.type() === 'error') {
+        logInfo(`Browser Console Error: "${msg.text()}"`);
+      }
+    });
```
Comment (aweng98) [inline playwright-e2e/fixtures/fixture-base.ts]: Write browser console error to test log.

Example

Screenshot 2024-01-30 at 2 17 45 PM

File: playwright-e2e/tests/dsm/kitUploadFlow/saliva-kit-prefix-check.spec.ts [hunk 13]
```diff
@@ -204,11 +204,28 @@ test.describe.serial('Saliva Kit Upload with a Canadian or New York address', ()
...
-        await kitsSentPage.waitForLoad();
+        await kitsSentPage.waitForReady();
         await kitsSentPage.selectKitType(kitType);
-        await kitsSentPage.search(KitsColumnsEnum.MF_CODE, kitLabel);
-        await kitsSentPage.assertDisplayedRowsCount(1);
+        await kitsSentPage.search(KitsColumnsEnum.MF_CODE, kitLabel, { count: 1 });
       });
     });
+
+    test(`Should throw error if same kit label and tracking label is used again @functional @dsm @${study}`, async ({page}) => {
```
Comment (aweng98) [inline playwright-e2e/tests/dsm/kitUploadFlow/saliva-kit-prefix-check.spec.ts:213]: New test that covers tracking scan error from duplicated scan.

File: playwright-e2e/utils/test-utils.ts [hunk 14]
```diff
@@ -40,22 +41,31 @@ export async function waitForNoSpinner(page: Page, opts: { timeout?: number } =
...
+    if (messageBody) {
+      response = await page.waitForResponse((response: Response) => response.url().includes(uri) &&
+      messageBody.every(async message => (await response.text()).includes(message as string)), { timeout });
+    } else {
+      response = await page.waitForResponse((response: Response) => response.url().includes(uri), { timeout });
+    }
+    await response.finished();
+  } catch (err) {
+    if (err instanceof errors.TimeoutError) {
+      throw new Error(`TimeoutError: waiting for URI: ${uri}: Timeout exceeded`);
```
Comment (aweng98) [inline playwright-e2e/utils/test-utils.ts:54]: better error in test log.

Example

Screenshot 2024-01-29 at 7 46 43 PM

File: playwright-e2e/fixtures/fixture-base.ts [hunk 15]
```diff
@@ -18,6 +19,12 @@ export const fixtureBase = base.extend({
     await page.route('**/*', async (route) => {
       return REQUEST_EXCLUDES.some((urlPart) => route.request().url().includes(urlPart)) ? route.abort() : route.continue();
     });
+    // Listen for all console events and log errors to the Terminal window
```
Comment (Rinchan20) [inline playwright-e2e/fixtures/fixture-base.ts:22]: nice!

File: playwright-e2e/utils/log-utils.ts [hunk 16]
```diff
@@ -1,44 +1,48 @@
...
+function localTime() {
+  return new Date().toLocaleTimeString();
+}
+
 export function logParticipantCreated(participantEmail: string, participantName: string) {
   test.info().annotations.push({
-    type: 'participant',
-    description: `${participantEmail} ${participantName}`
+    type: 'Participant',
+    description: `${localTime()}: ${participantEmail} ${participantName}`
```
Comment (Rinchan20) [inline playwright-e2e/utils/log-utils.ts:10]: nice!

**Review by Rinchan20 — APPROVED** [review]

**Comment by Rinchan20** [comment]
great changes, looking forward to improved logging

**Comment by aweng98** [comment]
Thank you. Update `.env` file after rebase.

**Pull request merged**